test/framework/src/main/java/org/elasticsearch/test/NodeRoles.java

```diff
@@ -0,0 +1,186 @@
+ * under the License.
+ */
+
+package org.elasticsearch.test;
+
+import org.elasticsearch.cluster.node.DiscoveryNodeRole;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.node.NodeRoleSettings;
+
+import java.util.List;
+import java.util.Set;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+/**
+ * Utility methods for creating {@link Settings} instances defining a set of {@link DiscoveryNodeRole}.
+ */
+public class NodeRoles {
+
+    public static Settings onlyRole(final DiscoveryNodeRole role) {
+        return onlyRole(Settings.EMPTY, role);
+    }
+
+    public static Settings onlyRole(final Settings settings, final DiscoveryNodeRole role) {
+        return onlyRoles(settings, Set.of(role));
+    }
+
+    public static Settings onlyRoles(final Set<DiscoveryNodeRole> roles) {
+        return onlyRoles(Settings.EMPTY, roles);
+    }
+
+    public static Settings onlyRoles(final Settings settings, final Set<DiscoveryNodeRole> roles) {
+        return Settings.builder()
+            .put(settings)
+            .putList(
+                NodeRoleSettings.NODE_ROLES_SETTING.getKey(),
+                roles.stream().map(DiscoveryNodeRole::roleName).collect(Collectors.toList()))
+            .build();
+    }
+
+    public static Settings removeRoles(final Set<DiscoveryNodeRole> roles) {
+        return removeRoles(Settings.EMPTY, roles);
+    }
+
+    public static Settings removeRoles(final Settings settings, final Set<DiscoveryNodeRole> roles) {
+        final Settings.Builder builder = Settings.builder().put(settings);
+        builder.putList(
+            NodeRoleSettings.NODE_ROLES_SETTING.getKey(),
+            NodeRoleSettings.NODE_ROLES_SETTING.get(settings)
+                .stream()
+                .filter(Predicate.not(roles::contains))
+                .map(DiscoveryNodeRole::roleName)
+                .collect(Collectors.toUnmodifiableList())
+        );
+        return builder.build();
+    }
+
+    public static Settings addRoles(final Set<DiscoveryNodeRole> roles) {
+        return addRoles(Settings.EMPTY, roles);
+    }
+
+    public static Settings addRoles(final Settings settings, final Set<DiscoveryNodeRole> roles) {
+        final Settings.Builder builder = Settings.builder().put(settings);
+        builder.putList(
+            NodeRoleSettings.NODE_ROLES_SETTING.getKey(),
+            NodeRoleSettings.NODE_ROLES_SETTING.get(settings)
+                .stream().filter(roles::contains)
+                .map(DiscoveryNodeRole::roleName)
+                .collect(Collectors.toUnmodifiableList())
+        );
+        return builder.build();
+    }
+
+    public static Settings noRoles() {
+        return noRoles(Settings.EMPTY);
+    }
+
+    public static Settings noRoles(final Settings settings) {
+        return Settings.builder().put(settings).putList(NodeRoleSettings.NODE_ROLES_SETTING.getKey(), List.of()).build();
```

Comment:
Would the intention be even clearer here if we used `Collections.emptyList`?

```suggestion
        return Settings.builder().put(settings).putList(NodeRoleSettings.NODE_ROLES_SETTING.getKey(), emptyList()).build();
```